### PR TITLE
rpc: Bumpfee units change, satoshis to BTC

### LINKED ISSUE #14815: bumpfee rpc uses satoshis as amount, not BTC
To the best of my knowledge all rpcs use BTC as amount, which means the json type is numeric or string. However `bumpfee` accepts an option that is denominated in satoshis:
```
     "totalFee"          (numeric, optional) Total fee (NOT feerate) to pay, in satoshis.
```

I am opening this issue to discuss how to avoid further inconsistencies like this, since they can easily lead to loosing a ton of money. (E.g. bumpfee with totalFee=1 is different from sendtoaddress with amount=1)

### PR DESCRIPTION
**Pull request by benthecarman**
Fixes #14815

### CODE REVIEW

**Comment by gmaxwell** [comment]
Operating on fee instead of feerate seems wrong to me in any case-- in particular bumping might change the transaction size in rather arbitrary ways.

File: test/functional/wallet_bumpfee.py [hunk 1]
```diff
@@ -207,17 +207,17 @@ def test_settxfee(rbf_node, dest_address):
...
+    bumped = rbf_node.bumpfee(rbfid, {"total_fee": 0.00002}) # 2000 sats
+    assert_raises_rpc_error(-4, "already bumped", rbf_node.bumpfee, rbfid, {"total_fee": 0.00003}) # 3000 sats
+    rbf_node.bumpfee(bumped["txid"], {"total_fee": 0.00003}) # 3000 sats
 
 
 def test_rebumping_not_replaceable(rbf_node, dest_address):
     # check that re-bumping a non-replaceable bump tx fails
     rbfid = spend_one_input(rbf_node, dest_address)
-    bumped = rbf_node.bumpfee(rbfid, {"totalFee": 10000, "replaceable": False})
+    bumped = rbf_node.bumpfee(rbfid, {"total_fee": 0.0001 , "replaceable": False})  # 10000 sats
```
Comment (practicalswift) [inline test/functional/wallet_bumpfee.py]: Drop space before comment :-)

**Comment by laanwj** [comment]
> Current implementation is breaking change. You should support the old option under the depreciation condition. It also needs release notes.

Concept ACK, but yes.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -3173,17 +3173,17 @@ static UniValue bumpfee(const JSONRPCRequest& request)
...
+                "Alternatively, the user can specify total_fee, or use RPC settxfee to set a higher fee rate.\n"
                 "At a minimum, the new fee rate must be high enough to pay an additional new relay fee (incrementalfee\n"
                 "returned by getnetworkinfo) to enter the node's mempool.\n",
                 {
                     {"txid", RPCArg::Type::STR_HEX, /* opt */ false, /* default_val */ "", "The txid to be bumped"},
                     {"options", RPCArg::Type::OBJ, /* opt */ true, /* default_val */ "null", "",
                         {
                             {"confTarget", RPCArg::Type::NUM, /* opt */ true, /* default_val */ "fallback to wallet's default", "Confirmation target (in blocks)"},
-                            {"totalFee", RPCArg::Type::NUM, /* opt */ true, /* default_val */ "fallback to 'confTarget'", "Total fee (NOT feerate) to pay, in satoshis.\n"
+                            {"total_fee", RPCArg::Type::NUM, /* opt */ true, /* default_val */ "fallback to 'confTarget'", "Total fee (NOT feerate) to pay, in bitcoins.\n"
```
Comment (MarcoFalke) [inline src/wallet/rpcwallet.cpp]: Type should be `AMOUNT` and not `NUM`

**Comment by benthecarman** [comment]
Rebased and fixed release notes

**Comment by practicalswift** [comment]
How can we make sure that an end-user who doesn't pay attention to the unit change isn't hurt financially?

**Comment by benthecarman** [comment]
> How can we make sure that an end-user who doesn't pay attention to the unit change isn't hurt financially?

I could add a sanity check to make sure the fee isn't over 1

**Comment by benthecarman** [comment]
Sanity check added

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -3241,43 +3292,82 @@ static UniValue bumpfee(const JSONRPCRequest& request)
...
+                true, true);
+
+            if (options.exists("confTarget") && options.exists("total_fee")) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "confTarget and total_fee options should not both be set. Please provide either a confirmation target for fee estimation or an explicit total fee for the transaction.");
+            } else if (options.exists("confTarget")) { // TODO: alias this to conf_target
+                coin_control.m_confirm_target = ParseConfirmTarget(options["confTarget"]);
+            } else if (options.exists("total_fee")) {
+                total_fee = AmountFromValue(options["total_fee"]); // Convert to satoshis
+                if (total_fee >= 100000000) { // User entered satoshis
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid total_fee, enter a value in bitcoins.");
```
Comment (practicalswift) [inline src/wallet/rpcwallet.cpp]: The error should include why the `total_fee` is invalid (as done below).

Also this error condition should be tested :-)

**Comment by luke-jr** [comment]
Does this really need to duplicate all the code?

**Comment by benthecarman** [comment]
> Does this really need to duplicate all the code?

It doesn't, but I did that so it'd be easier to remove the depreciated version in the future.

**Comment by luke-jr** [comment]
I don't think it makes it any easier, just harder to review the PR... :/

**Comment by benthecarman** [comment]
Alright I'll clean it up then

**Comment by benthecarman** [comment]
Should be ready for a re-review

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -3259,34 +3262,42 @@ static UniValue bumpfee(const JSONRPCRequest& request)
...
+        if (options.exists("confTarget") && (options.exists(totalFeeName))) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "confTarget and " + totalFeeName +" options should not both be set. Please provide either a confirmation target for fee estimation or an explicit total fee for the transaction.");
         } else if (options.exists("confTarget")) { // TODO: alias this to conf_target
             coin_control.m_confirm_target = ParseConfirmTarget(options["confTarget"], pwallet->chain().estimateMaxBlocks());
-        } else if (options.exists("totalFee")) {
-            totalFee = options["totalFee"].get_int64();
-            if (totalFee <= 0) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid totalFee %s (must be greater than 0)", FormatMoney(totalFee)));
+        } else if (options.exists("total_fee") && !IsDeprecatedRPCEnabled("bumpfee")) {
+            total_fee = AmountFromValue(options["total_fee"]); // Convert to satohis
```
Comment (practicalswift) [inline src/wallet/rpcwallet.cpp]: Should be "satoshis" :-)

**Comment by luke-jr** [comment]
It should probably remain possible to use the new API even when the deprecated one is enabled...

**Comment by benthecarman** [comment]
Going to close this issue, seems the totalFee argument is being deprecated in #15996

**Pull request closed without merging**